### PR TITLE
Feature/loans tests

### PR DESCRIPTION
**Pull request by simplemachine92**
# Description

*What does this PR: do, how, why?*

## Limitations & risks

*Are there any trade-off or new vulnarbility surface based on theses changes?*

# Check-list
- [ ] Tests are covering the new feature
- [ ] Code is [natspec'd](https://docs.soliditylang.org/en/v0.8.17/natspec-format.html)
- [ ] Code is [linted and formatted](https://docs.soliditylang.org/en/v0.8.17/style-guide.html)
- [ ] I have run the test locally (and they pass)
- [ ] I have rebased to the latest main commit (and tests still pass)

# Interactions
These changes will impact the following contracts:
- Directly:

- Indirectly:

### CODE REVIEW

File: src/REVLoans.sol [hunk 1]
```diff
@@ -308,43 +314,31 @@ contract REVLoans is ERC721, ERC2771Context, IREVLoans {
         external
         payable
         override
-        returns (uint256 newLoanId, REVLoan memory, REVLoan memory newLoan)
+        returns (uint256 refinancedLoanId, uint256 newLoanId, REVLoan memory, REVLoan memory newLoan)
```
Comment (xBA5ED) [inline src/REVLoans.sol:317]: If we need `nonReentrant` on the other public functions we should add it here as well to be safe. Perhaps we can make `borrowFrom(..)` an external function that has the modifier and copy all its current logic to `_borrowFrom(..)` which we can also use here.

File: src/REVLoans.sol [hunk 2]
```diff
@@ -354,46 +348,38 @@ contract REVLoans is ERC721, ERC2771Context, IREVLoans {
...
 
+        // Keep a reference to the fee that'll be taken.
+        uint256 sourceFeeAmount = _determineSourceFeeAmount(loan, amount);
+
+        // Make sure the amount being paid off is less than the loan's amount.
+        if (amount - sourceFeeAmount > loan.amount) revert AMOUNT_EXCEEDS_LOAN();
+
         // If the amount being paid is greater than the loan's amount, return extra to the payer.
         if (amount > loan.amount + sourceFeeAmount) {
             _transferFrom({
```
Comment (xBA5ED) [inline src/REVLoans.sol:385]: Don't these two contradict each other?

Comment (mejango) [reply src/REVLoans.sol]: nice. i left the transferFrom, and reset `amount -= (sourceFeeAmount + loan.amount)`

File: src/REVLoans.sol (hunk 1 above)
Comment (mejango) [reply src/REVLoans.sol:316]: true. im not yet convinced reentrency is needed cc @simplemachine92 do we have a proof?

Comment (simplemachine92) [reply src/REVLoans.sol:316]: it was precautionary, but I'll confirm/deny in tests.

File: src/REVLoans.sol [hunk 3]
```diff
@@ -391,6 +392,8 @@ contract REVLoans is ERC721, ERC2771Context, IREVLoans, ReentrancyGuard {
 
         if (collateralToReturn > loan.collateral) revert COLLATERAL_EXCEEDS_LOAN();
 
+        if (amount == 0) revert ZERO_AMOUNT();
```
Comment (simplemachine92) [inline src/REVLoans.sol:395]: ah, sorry, this should probably be after 398?

**Pull request merged**